text/0340-firehose-l2.md

```diff
@@ -0,0 +1,944 @@
+not being delivered to your destination), or when the rate of incoming records exceeds the
+limit of records per second (indicating data is flowing into your delivery stream faster
+than it is configured to process).
+
+CDK provides methods for accessing delivery stream metrics with default configuration,
+such as `metricIncomingBytes`, and `metricIncomingRecords` (see [`IDeliveryStream`](../lib/delivery-stream.ts)
+for a full list). CDK also provides a generic `metric` method that can be used to produce
+metric configurations for any metric provided by Kinesis Data Firehose; the configurations
+are pre-populated with the correct dimensions for the delivery stream.
+
+```ts fixture=with-delivery-stream
+// TODO: confirm this is a valid alarm
+import * as cloudwatch from '@aws-cdk/aws-cloudwatch';
+// Alarm that triggers when the per-second average of incoming bytes exceeds 90% of the current service limit
+const incomingBytesPercentOfLimit = new cloudwatch.MathExpression({
+  expression: 'incomingBytes / 300 / bytePerSecLimit',
+  usingMetrics: {
+    incomingBytes: deliveryStream.metricIncomingBytes({ statistic: cloudwatch.Statistic.SUM }),
+    bytePerSecLimit: deliveryStream.metric('BytesPerSecondLimit'),
+  },
+});
+new Alarm(this, 'Alarm', {
+  metric: incomingBytesPercentOfLimit,
+  threshold: 0.9,
+  evaluationPeriods: 3,
+});
+```
+
+See: [Monitoring Using CloudWatch Metrics](https://docs.aws.amazon.com/firehose/latest/dev/monitoring-with-cloudwatch-metrics.html)
+in the *Kinesis Data Firehose Developer Guide*.
+
+## Compression
+
+Your data can automatically be compressed when it is delivered to S3 as either a final or
+an intermediary/backup destination. Supported compression formats are: gzip, Snappy,
+Hadoop-compatible Snappy, and ZIP, except for Redshift destinations, where Snappy
+(regardless of Hadoop-compatibility) and ZIP are not supported. By default, data is
+delivered to S3 without compression.
+
+```ts fixture=with-bucket
+// Compress data delivered to S3 using Snappy
+const s3Destination = new destinations.S3({
+  compression: Compression.SNAPPY,
+  bucket: bucket,
+});
+new DeliveryStream(this, 'Delivery Stream', {
+  destination: destination,
+});
+```
+
+## Buffering
+
+Incoming data is buffered before it is delivered to the specified destination. The
+delivery stream will wait until the amount of incoming data has exceeded some threshold
+(the "buffer size") or until the time since the last data delivery occurred exceeds some
+threshold (the "buffer interval"), whichever happens first. You can configure these
+thresholds based on the capabilities of the destination and your use-case. By default, the
+buffer size is 3 MiB and the buffer interval is 1 minute.
+
+```ts fixture=with-bucket
+// Increase the buffer interval and size to 5 minutes and 3 MiB, respectively
+import * as cdk from '@aws-cdk/core';
+
+const s3Destination = new destinations.S3({
+  bufferingInterval: cdk.Duration.minutes(5),
```

Comment:
The bufferingInterval and bufferSize are not supported for all destinations. It's supported for all destinations except Redshift and Splunk.

So we could make this a part of the deliverystream, but then would need to have a synth-time error for the destinations that don't support it. Or, it would silently do nothing for those destinations.

Reply:
Redshift does in fact support buffering via the intermediate S3 bucket. As in, data is buffered by Firehose before being delivered to S3 and then is immediately COPY'd to the Redshift cluster

Reply:
Since we will be supporting multiple destinations, this should not be moved to the delivery stream